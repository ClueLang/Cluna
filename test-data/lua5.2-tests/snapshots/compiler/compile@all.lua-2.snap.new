---
source: src/main.rs
assertion_line: 118
expression: compiled
input_file: test-data/lua5.2-tests/all.lua
---
local version = "Lua 5.2"
if _VERSION ~= version {
    io.stderr::write("\nThis test suite is for ", version, ", not for ", _VERSION, "\nExiting tests\n");
    return 
}
_soft = false
_port = false
_no32 = false
_nomsg = false
_noposix = false
_nolonglong = false
_noformatA = false
local usertests = rawget(_G, "_U")
if usertests {
    _soft = true
    _port = true
    _no32 = true
    _nomsg = true
    _noposix = true
    _nolonglong = true
    _noformatA = true
}
if usertests {
    T = nil
}
T = rawget(_G, "T")
package.path = "?;./?.lua" .. package.path
math.randomseed(0);
collectgarbage("setstepmul", 200);
collectgarbage("setpause", 200);
print("current path:\n****" .. package.path .. "****\n");
local c = os.clock()
local collectgarbage = collectgarbage
{
    local msgs = {}
    global fn Message(m) {
        if !_nomsg {
            print(m);
            msgs[(#msgs + 1)] = string.sub(m, 3,  - 3)
        }
    }
    assert(os.setlocale("C"));
    local T, print, format, write, assert, type, unpack, floor = T, print, string.format, io.write, assert, type, table.unpack, math.floor
    local fn F(m) {
        local fn round(m) {
            m = m + 0.04999
            return m - (m % 0.1)
        }
        if m < 1000 {
            return m
        } else {
            m = m / 1000
            if m < 1000 {
                return round(m) .. "K"
            } else {
                return round(m / 1000) .. "M"
            }
        }
    }
    local showmem
    if !T {
        local max = 0
        showmem = fn () {
            local m = collectgarbage("count") * 1024
            max = (m > max) && m || max
            print(format("    ---- total memory: %s, max memory: %s ----\n", F(m), F(max)));
        }
    } else {
        showmem = fn () {
            T.checkmemory();
            local total, numblocks, maxmem = T.totalmem()
            local count = collectgarbage("count")
            print(format("\n    ---- total memory: %s (%.0fK), max use: %s,  blocks: %d\n", F(total), count, F(maxmem), numblocks));
            print(format("\t(strings:  %d, tables: %d, functions: %d, " .. "\n\tudata: %d, threads: %d)", T.totalmem("string"), T.totalmem("table"), T.totalmem("function"), T.totalmem("userdata"), T.totalmem("thread")));
        }
    }
    local fn report(n) {
        print("\n***** FILE '" .. n .. "'*****");
    }
    local olddofile = dofile
    dofile = fn (n) {
        showmem();
        report(n);
        local f = assert(loadfile(n))
        local b = string.dump(f)
        f = assert(load(b))
        return f()
    }
    dofile('main.lua');
    {
        local next, setmetatable, stderr = next, setmetatable, io.stderr
        local mt = {}
        mt.__gc = fn (o) {
            stderr::write('.');
            local n = setmetatable({}, mt)
            o = nil
            local a, b, c, d, e = nil
        }
        local n = setmetatable({}, mt)
    }
    report("gc.lua");
    local f = assert(loadfile('gc.lua'))
    f();
    collectgarbage("generational");
    dofile('db.lua');
    assert(dofile('calls.lua') == deep && deep);
    olddofile('strings.lua');
    olddofile('literals.lua');
    assert(dofile('attrib.lua') == 27);
    collectgarbage("incremental");
    olddofile('strings.lua');
    olddofile('literals.lua');
    dofile('constructs.lua');
    dofile('api.lua');
    collectgarbage("generational");
    collectgarbage("setpause", 200);
    collectgarbage("setmajorinc", 500);
    assert(dofile('locals.lua') == 5);
    dofile('constructs.lua');
    dofile('code.lua');
    if !_G._soft {
        report('big.lua');
        local f = coroutine.wrap(assert(loadfile('big.lua')))
        assert(f() == 'b');
        assert(f() == 'a');
    }
    dofile('nextvar.lua');
    dofile('pm.lua');
    dofile('api.lua');
    assert(dofile('events.lua') == 12);
    dofile('vararg.lua');
    dofile('closure.lua');
    dofile('coroutine.lua');
    dofile('goto.lua');
    dofile('errors.lua');
    dofile('math.lua');
    dofile('sort.lua');
    dofile('bitwise.lua');
    assert(dofile('verybig.lua') == 10);
    collectgarbage();
    dofile('files.lua');
    if #msgs > 0 {
        print("\ntests not performed:");
        for i = 1, #msgs {
            print(msgs[(i)]);
        }
        print();
    }
    print("final OK !!!");
    local debug = require("debug")
    debug.sethook(fn (a) {
        assert(type(a) == 'string');
    }, "cr");
    _G.showmem = showmem
}
local _G, showmem, print, format, clock, assert, open = _G, showmem, print, string.format, os.clock, assert, io.open
local fname = T && "time-debug.txt" || "time.txt"
local lasttime
if !usertests {
    local f = io.open(fname)
    if f {
        lasttime = assert(tonumber(f::read('*a')))
        f::close();
    } else {
        lasttime = nil
    }
}
print('cleaning all!!!!');
for n with pairs(_G) {
    if !({
        ___Glob = 1, 
        tostring = 1
    })[(n)] {
        _G[(n)] = nil
    }
}
collectgarbage();
collectgarbage();
collectgarbage();
collectgarbage();
collectgarbage();
collectgarbage();
showmem();
local time = clock() - c
print(format("\n\ntotal time: %.2f\n", time));
if !usertests {
    lasttime = lasttime || time
    local diff = (time - lasttime) / time
    local tolerance = 0.05
    assert(diff < tolerance && diff >  - tolerance);
    assert(open(fname, "w"))::write(time)::close();
}
