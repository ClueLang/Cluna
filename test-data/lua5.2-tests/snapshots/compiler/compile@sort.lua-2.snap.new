---
source: src/main.rs
assertion_line: 118
expression: compiled
input_file: test-data/lua5.2-tests/sort.lua
---
print("testing (parts of) table library");
print("testing unpack");
local unpack = table.unpack
local x, y, z, a, n
a = {}
lim = 2000
for i = 1, lim {
    a[(i)] = i
}
assert(select(lim, unpack(a)) == lim && select('#', unpack(a)) == lim);
x = unpack(a)
assert(x == 1);
x = {
    unpack(a)
}
assert(#x == lim && x[(1)] == 1 && x[(lim)] == lim);
x = {
    unpack(a, lim - 2)
}
assert(#x == 3 && x[(1)] == lim - 2 && x[(3)] == lim);
x = {
    unpack(a, 10, 6)
}
assert(next(x) == nil);
x = {
    unpack(a, 11, 10)
}
assert(next(x) == nil);
x, y = unpack(a, 10, 10)
assert(x == 10 && y == nil);
x, y, z = unpack(a, 10, 11)
assert(x == 10 && y == 11 && z == nil);
a, x = unpack({
    1
})
assert(a == 1 && x == nil);
a, x = unpack({
    1, 
    2
}, 1, 1)
assert(a == 1 && x == nil);
if !_no32 {
    assert(!pcall(unpack, {}, 0, 2 ^ 31 - 1));
    assert(!pcall(unpack, {}, 1, 2 ^ 31 - 1));
    assert(!pcall(unpack, {},  - (2 ^ 31), 2 ^ 31 - 1));
    assert(!pcall(unpack, {},  - (2 ^ 31 - 1), 2 ^ 31 - 1));
    assert(pcall(unpack, {}, 2 ^ 31 - 1, 0));
    assert(pcall(unpack, {}, 2 ^ 31 - 1, 1));
    pcall(unpack, {}, 1, 2 ^ 31);
    a, b = unpack({
        2 ^ 31 - 1 = 20
    }, 2 ^ 31 - 1, 2 ^ 31 - 1)
    assert(a == 20 && b == nil);
    a, b = unpack({
        2 ^ 31 - 1 = 20
    }, 2 ^ 31 - 2, 2 ^ 31 - 1)
    assert(a == nil && b == 20);
}
print("testing pack");
a = table.pack()
assert(a[(1)] == nil && a.n == 0);
a = table.pack(table)
assert(a[(1)] == table && a.n == 1);
a = table.pack(nil, nil, nil, nil)
assert(a[(1)] == nil && a.n == 4);
print("testing sort");
local fn check(t) {
    local fn f(a, b) {
        assert(a && b);
        return true
    }
    local s, e = pcall(table.sort, t, f)
    assert(!s && e::find("invalid order function"));
}
check({
    1, 
    2, 
    3, 
    4
});
check({
    1, 
    2, 
    3, 
    4, 
    5
});
check({
    1, 
    2, 
    3, 
    4, 
    5, 
    6
});
global fn check(a, f) {
    f = f || fn (x, y) {
        return x < y
    }
    for n = #a, 2,  - 1 {
        assert(!f(a[(n)], a[(n - 1)]));
    }
}
a = {
    "Jan", 
    "Feb", 
    "Mar", 
    "Apr", 
    "May", 
    "Jun", 
    "Jul", 
    "Aug", 
    "Sep", 
    "Oct", 
    "Nov", 
    "Dec"
}
table.sort(a);
check(a);
global fn perm(s, n) {
    n = n || #s
    if n == 1 {
        local t = {
            unpack(s)
        }
        table.sort(t);
        check(t);
    } else {
        for i = 1, n {
            s[(i)], s[(n)] = s[(n)], s[(i)]
            perm(s, n - 1);
            s[(i)], s[(n)] = s[(n)], s[(i)]
        }
    }
}
perm({});
perm({
    1
});
perm({
    1, 
    2
});
perm({
    1, 
    2, 
    3
});
perm({
    1, 
    2, 
    3, 
    4
});
perm({
    2, 
    2, 
    3, 
    4
});
perm({
    1, 
    2, 
    3, 
    4, 
    5
});
perm({
    1, 
    2, 
    3, 
    3, 
    5
});
perm({
    1, 
    2, 
    3, 
    4, 
    5, 
    6
});
perm({
    2, 
    2, 
    3, 
    3, 
    5, 
    6
});
limit = 30000
if _soft {
    limit = 5000
}
a = {}
for i = 1, limit {
    a[(i)] = math.random()
}
local x = os.clock()
table.sort(a);
print(string.format("Sorting %d elements in %.2f sec.", limit, os.clock() - x));
check(a);
x = os.clock()
table.sort(a);
print(string.format("Re-sorting %d elements in %.2f sec.", limit, os.clock() - x));
check(a);
a = {}
for i = 1, limit {
    a[(i)] = math.random()
}
x = os.clock()
i = 0
table.sort(a, fn (x, y) {
    i = i + 1
    return y < x
});
print(string.format("Invert-sorting other %d elements in %.2f sec., with %i comparisons", limit, os.clock() - x, i));
check(a, fn (x, y) {
    return y < x
});
table.sort({});
for i = 1, limit {
    a[(i)] = false
}
x = os.clock()
table.sort(a, fn (x, y) {
    return nil
});
print(string.format("Sorting %d equal elements in %.2f sec.", limit, os.clock() - x));
check(a, fn (x, y) {
    return nil
});
for i, v with pairs(a) {
    assert(!v || i == 'n' && v == limit);
}
A = {
    "Ã¡lo", 
    "\0first :-)", 
    "alo", 
    "then this one", 
    "45", 
    "and a new"
}
table.sort(A);
check(A);
table.sort(A, fn (x, y) {
    load(string.format("A[%q] = ''", x))();
    collectgarbage();
    return x < y
});
tt = {
    __lt = fn (a, b) {
        return a.val < b.val
    }
}
a = {}
for i = 1, 10 {
    a[(i)] = {
        val = math.random(100)
    }
    setmetatable(a[(i)], tt);
}
table.sort(a);
check(a, tt.__lt);
check(a);
print("OK");
