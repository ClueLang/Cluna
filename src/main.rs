#![allow(non_snake_case)]
#![allow(non_upper_case_globals)]

macro_rules! check {
    ($tocheck: expr) => {
        match $tocheck {
            Ok(t) => t,
            Err(e) => return Err(e.to_string()),
        }
    };
}

macro_rules! arg {
    ($name: expr) => {
        unsafe { $name }
    };
}

// mod compiler;
mod compiler;
mod parser;
mod scanner;

use clap::Parser;
use compiler::compile_tokens;
use parser::*;
use scanner::*;
use std::{fs, fs::File, io::prelude::*, path::Path, time::Instant};

pub static mut ENV_TOKENS: bool = false;
pub static mut ENV_STRUCT: bool = false;
pub static mut ENV_OUTPUT: bool = false;
pub static mut ENV_DONTSAVE: bool = false;
pub static mut ENV_PATHISCODE: bool = false;
pub static mut ENV_NOMULTILINE: bool = false;

#[derive(Parser)]
#[clap(about, version, long_about = None)]
struct Cli {
    /// The path to the directory where the *.lua files are located.
    /// Every directory inside the given directory will be checked too.
    /// If the path points to a single *.lua file, only that file will be compiled.
    #[clap(required_unless_present = "license")]
    path: Option<String>,

    /// The name the output file will have
    #[clap(default_value = "main", value_name = "OUTPUT FILE NAME")]
    outputname: String,

    /// Print license information
    #[clap(short = 'L', long, display_order = 1000)]
    license: bool,

    /// Print list of detected tokens in compiled files
    #[clap(long)]
    tokens: bool,

    /// Print syntax structure of the tokens of the compiled files
    #[clap(long)]
    r#struct: bool,

    /// Print output Clue code in the console
    #[clap(short, long)]
    output: bool,

    /// Don't save compiled code
    #[clap(short = 'D', long)]
    dontsave: bool,

    /// Treat PATH not as a path but as clue code
    #[clap(short, long)]
    pathiscode: bool,

    /// Don't use multiline strings in the output
    #[clap(long)]
    nomultiline: bool,
}

fn compile_code(code: String, name: String, scope: usize) -> Result<String, String> {
    let time = Instant::now();
    let (tokens, _comments): (Vec<Token>, Vec<Comment>) = scan_code(code, name.clone())?;
    if arg!(ENV_TOKENS) {
        println!("Scanned tokens of file \"{}\":\n{:#?}", name, tokens);
    }
    let ctokens = parse_tokens(tokens, name.clone())?;
    if arg!(ENV_STRUCT) {
        println!("Parsed structure of file \"{}\":\n{:#?}", name, ctokens);
    }

    let code = compile_tokens(scope, ctokens);
    if arg!(ENV_OUTPUT) {
        println!("Compiled Lua code of file \"{}\":\n{}", name, code);
    }
    println!(
        "Compiled file \"{}\" in {} seconds!",
        name,
        time.elapsed().as_secs_f32()
    );
    Ok(code)
}

fn compile_file(path: &Path, name: String, scope: usize) -> Result<String, String> {
    let mut code: String = String::new();
    check!(check!(File::open(path)).read_to_string(&mut code));
    Ok(compile_code(code, name, scope)?)
}

fn compile_folder(path: &Path, rpath: String) -> Result<(), String> {
    for entry in check!(fs::read_dir(path)) {
        let entry = check!(entry);
        let name: String = entry
            .path()
            .file_name()
            .unwrap()
            .to_string_lossy()
            .into_owned();
        let filePathName: String = format!("{}/{}", path.display(), name);
        let filepath: &Path = &Path::new(&filePathName);
        let rname = rpath.clone() + &name;
        if filepath.is_dir() {
            compile_folder(filepath, rname + ".")?;
        } else if filePathName.ends_with(".lua") {
            let code = compile_file(filepath, name, 2)?;
            let compiled_path = format!(
                "{}{}",
                filepath.display().to_string().strip_suffix(".lua").unwrap(),
                ".clue"
            );
            if !arg!(ENV_DONTSAVE) {
                check!(fs::write(compiled_path, code))
            }
        }
    }
    Ok(())
}

fn main() -> Result<(), String> {
    let cli = Cli::parse();
    if cli.license {
        println!("{}", include_str!("../LICENSE"));
        return Ok(());
    }
    unsafe {
        ENV_TOKENS = cli.tokens;
        ENV_STRUCT = cli.r#struct;
        ENV_OUTPUT = cli.output;
        ENV_DONTSAVE = cli.dontsave;
        ENV_PATHISCODE = cli.pathiscode;
        ENV_NOMULTILINE = cli.nomultiline;
    }
    let codepath = cli.path.unwrap();
    if arg!(ENV_PATHISCODE) {
        println!(
            "{}",
            compile_code(codepath.clone(), "(command line)".to_owned(), 0)?
        );
        return Ok(());
    }
    let path: &Path = Path::new(&codepath);

    if path.is_dir() {
        compile_folder(path, String::new())?;
    } else if path.is_file() {
        let compiled_name =
            String::from(path.display().to_string().strip_suffix(".lua").unwrap()) + ".clue";
        let code = compile_file(
            path,
            path.file_name().unwrap().to_string_lossy().into_owned(),
            0,
        )?;

        if !arg!(ENV_DONTSAVE) {
            check!(fs::write(compiled_name, code))
        }
    } else {
        return Err(String::from("The given path doesn't exist"));
    }
    Ok(())
}

#[cfg(test)]
mod test {
    use crate::compile_folder;
    use std::path::Path;

    #[test]
    fn compile() {
        compile_folder(Path::new("tests/"), String::new()).unwrap();
    }
}
