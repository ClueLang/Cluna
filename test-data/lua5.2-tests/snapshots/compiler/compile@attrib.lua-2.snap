---
source: src/main.rs
expression: compiled
input_file: test-data/lua5.2-tests/attrib.lua
---
if !_port {
    print("testing require");
    assert(require("string") == string);
    assert(require("math") == math);
    assert(require("table") == table);
    assert(require("io") == io);
    assert(require("os") == os);
    assert(require("coroutine") == coroutine);
    assert(type(package.path) == "string");
    assert(type(package.cpath) == "string");
    assert(type(package.loaded) == "table");
    assert(type(package.preload) == "table");
    assert(type(package.config) == "string");
    print("package config: " .. string.gsub(package.config, "\n", "|"));
    {
        local max = 2000
        local t = {}
        for i = 1, max {
            t[(i)] = string.rep("?", i % 10 + 1)
        }
        t[(#t + 1)] = ";"
        local path = table.concat(t, ";")
        local s, err = package.searchpath("xuxu", path)
        assert(!s && string.find(err, string.rep("xuxu", 10)) && #string.gsub(err, "[^\n]", "") >= max);
        local path = string.rep("?", max)
        local s, err = package.searchpath("xuxu", path)
        assert(!s && string.find(err, string.rep('xuxu', max)));
    }
    {
        local oldpath = package.path
        package.path = {}
        local s, err = pcall(require, "no-such-file")
        assert(!s && string.find(err, "package.path"));
        package.path = oldpath
    }
    print('+');
    local DIR = "libs/"
    local fn D(x) {
        return DIR .. x
    }
    local fn createfiles(files, preextras, posextras) {
        for n, c with pairs(files) {
            io.output(D(n));
            io.write(string.format(preextras, n));
            io.write(c);
            io.write(string.format(posextras, n));
            io.close(io.output());
        }
    }
    global fn removefiles(files) {
        for n with pairs(files) {
            os.remove(D(n));
        }
    }
    local files = {
        "names.lua" = "do return {...} end\n", 
        "err.lua" = "B = 15; a = a + 1;", 
        "A.lua" = "", 
        "B.lua" = "assert(...=='B');require 'A'", 
        "A.lc" = "", 
        "A" = "", 
        "L" = "", 
        "XXxX" = "", 
        "C.lua" = "package.loaded[...] = 25; require'C'"
    }
    AA = nil
    local extras = `
NAME = '%s'
REQUIRED = ...
return AA`
    createfiles(files, "", extras);
    assert(package.searchpath("C.lua", D("?"), "", "") == D("C.lua"));
    assert(package.searchpath("C.lua", D("?"), ".", ".") == D("C.lua"));
    assert(package.searchpath("--x-", D("?"), "-", "X") == D("XXxX"));
    assert(package.searchpath("---xX", D("?"), "---", "XX") == D("XXxX"));
    assert(package.searchpath(D("C.lua"), "?", "/") == D("C.lua"));
    assert(package.searchpath(".\\C.lua", D("?"), "\\") == D("./C.lua"));
    local oldpath = package.path
    package.path = string.gsub("D/?.lua;D/?.lc;D/?;D/??x?;D/L", "D/", DIR)
    local try = fn (p, n, r) {
        NAME = nil
        local rr = require(p)
        assert(NAME == n);
        assert(REQUIRED == p);
        assert(rr == r);
    }
    a = require("names")
    assert(a[(1)] == "names" && a[(2)] == D("names.lua"));
    _G.a = nil
    assert(!pcall(require, "err"));
    assert(B == 15);
    assert(package.searchpath("C", package.path) == D("C.lua"));
    assert(require("C") == 25);
    assert(require("C") == 25);
    AA = nil
    try('B', 'B.lua', true);
    assert(package.loaded.B);
    assert(require("B") == true);
    assert(package.loaded.A);
    assert(require("C") == 25);
    package.loaded.A = nil
    try('B', nil, true);
    try('A', 'A.lua', true);
    package.loaded.A = nil
    os.remove(D('A.lua'));
    AA = {}
    try('A', 'A.lc', AA);
    assert(package.searchpath("A", package.path) == D("A.lc"));
    assert(require("A") == AA);
    AA = false
    try('K', 'L', false);
    try('K', 'L', false);
    assert(rawget(_G, "_REQUIREDNAME") == nil);
    AA = "x"
    try("X", "XXxX", AA);
    removefiles(files);
    local _G = _G
    package.path = string.gsub("D/?.lua;D/?/init.lua", "D/", DIR)
    files = {
        "P1/init.lua" = "AA = 10", 
        "P1/xuxu.lua" = "AA = 20"
    }
    createfiles(files, "_ENV = {}\n", "\nreturn _ENV\n");
    AA = 0
    local m = assert(require("P1"))
    assert(AA == 0 && m.AA == 10);
    assert(require("P1") == m);
    assert(require("P1") == m);
    assert(package.searchpath("P1.xuxu", package.path) == D("P1/xuxu.lua"));
    m.xuxu = assert(require("P1.xuxu"))
    assert(AA == 0 && m.xuxu.AA == 20);
    assert(require("P1.xuxu") == m.xuxu);
    assert(require("P1.xuxu") == m.xuxu);
    assert(require("P1") == m && m.AA == 10);
    removefiles(files);
    package.path = ""
    assert(!pcall(require, "file_does_not_exist"));
    package.path = "??\0?"
    assert(!pcall(require, "file_does_not_exist1"));
    package.path = oldpath
    local fname = "file_does_not_exist2"
    local m, err = pcall(require, fname)
    for t with string.gmatch(package.path .. ";" .. package.cpath, "[^;]+") {
        t = string.gsub(t, "?", fname)
        assert(string.find(err, t, 1, true));
    }
    local fn import(...) {
        local f = {
            ...
        }
        return fn (m) {
            for i = 1, #f {
                m[(f[(i)])] = _G[(f[(i)])]
            }
        }
    }
    assert(!pcall(module, 'XUXU'));
    local p = ""
    local st, err, when = package.loadlib(D("lib1.so"), "*")
    if !st {
        local f, err, when = package.loadlib("donotexist", p .. "xuxu")
        assert(!f && type(err) == "string" && when == "absent");
        (Message || print)('\a\n >>> cannot load dynamic library <<<\n\a');
        print(err, when);
    } else {
        local f = assert(package.loadlib(D("lib1.so"), p .. "onefunction"))
        local a, b = f(15, 25)
        assert(a == 25 && b == 15);
        f = assert(package.loadlib(D("lib1.so"), p .. "anotherfunc"))
        assert(f(10, 20) == "1020\n");
        local f, err, when = package.loadlib(D("lib1.so"), p .. "xuxu")
        assert(!f && type(err) == "string" && when == "init");
        f, err, when = package.loadlib("donotexist", p .. "xuxu")
        assert(!f && type(err) == "string" && when == "open");
        f = assert(package.loadlib(D("lib11.so"), p .. "luaopen_lib11"))
        assert(f() == "exported");
        package.cpath = D("?.so")
        local lib2 = require("v-lib2")
        assert(_ENV.x == "v-lib2" && _ENV.y == D("v-lib2.so"));
        assert(lib2.id("x") == "x");
        local fs = require("lib1.sub")
        assert(_ENV.x == "lib1.sub" && _ENV.y == D("lib1.so"));
        assert(fs.id(45) == 45);
    }
    _ENV = _G
    {
        local p = package
        package = {}
        p.preload.pl = fn (...) {
            local _ENV = {
                ...
            }
            global fn xuxu(x) {
                return x + 20
            }
            return _ENV
        }
        local pl = require("pl")
        assert(require("pl") == pl);
        assert(pl.xuxu(10) == 30);
        assert(pl[(1)] == "pl" && pl[(2)] == nil);
        package = p
        assert(type(package.path) == "string");
    }
    print('+');
}
print("testing assignments, logical operators, and constructors");
local res, res2 = 27
a, b = 1, 2 + 3
assert(a == 1 && b == 5);
a = {}
global fn f() {
    return 10, 11, 12
}
a.x, b, a[(1)] = 1, 2, f()
assert(a.x == 1 && b == 2 && a[(1)] == 10);
a[(f())], b, a[(f() + 3)] = f(), a, 'x'
assert(a[(10)] == 10 && b == a && a[(13)] == 'x');
{
    local f = fn (n) {
        local x = {}
        for i = 1, n {
            x[(i)] = i
        }
        return table.unpack(x)
    }
    local a, b, c
    a, b = 0, f(1)
    assert(a == 0 && b == 1);
    A, b = 0, f(1)
    assert(A == 0 && b == 1);
    a, b, c = 0, 5, f(4)
    assert(a == 0 && b == 5 && c == 1);
    a, b, c = 0, 5, f(0)
    assert(a == 0 && b == 5 && c == nil);
}
a, b, c, d = 1 && nil, 1 || nil, (1 && (nil || 1)), 6
assert(!a && b && c && d == 6);
d = 20
a, b, c, d = f()
assert(a == 10 && b == 11 && c == 12 && d == nil);
a, b = f(), 1, 2, 3, f()
assert(a == 10 && b == 1);
assert(a < b == false && a > b == true);
assert((10 && 2) == 2);
assert((10 || 2) == 10);
assert((10 || assert(nil)) == 10);
assert(!(nil && assert(nil)));
assert((nil || "alo") == "alo");
assert((nil && 10) == nil);
assert((false && 10) == false);
assert((true || 10) == true);
assert((false || 10) == 10);
assert(false != nil);
assert(nil != false);
assert(!nil == true);
assert(!!nil == false);
assert(!!1 == true);
assert(!!a == true);
assert(!!(6 || nil) == true);
assert(!!(nil && 56) == false);
assert(!!(nil && true) == false);
assert({} != {});
print('+');
a = {}
a[(true)] = 20
a[(false)] = 10
assert(a[(1 < 2)] == 20 && a[(1 > 2)] == 10);
global fn f(a) {
    return a
}
local a = {}
for i = 3000, -3000, -1 {
    a[(i)] = i
}
a[(10e30)] = "alo"
a[(true)] = 10
a[(false)] = 20
assert(a[(10e30)] == 'alo' && a[(!1)] == 20 && a[(10 < 20)] == 10);
for i = 3000, -3000, -1 {
    assert(a[(i)] == i);
}
a[(print)] = assert
a[(f)] = print
a[(a)] = a
assert(a[(a)][(a)][(a)][(a)][(print)] == assert);
a[(print)](a[(a[(f)])] == a[(print)]);
assert(!pcall(fn () {
    local a = {}
    a[(nil)] = 10
}));
assert(!pcall(fn () {
    local a = {
        nil = 10
    }
}));
assert(a[(nil)] == nil);
a = nil
a = {
    10, 
    9, 
    8, 
    7, 
    6, 
    5, 
    4, 
    3, 
    2, 
    -3 = 'a', 
    f = print, 
    a = 'a', 
    b = 'ab'
}
a, a.x, a.y = a, a[(-3)]
assert(a[(1)] == 10 && a[(-3)] == a.a && a[(f)] == print && a.x == 'a' && !a.y);
a[(1)], f(a)[(2)], b, c = {
    'alo' = assert
}, 10, a[(1)], a[(f)], 6, 10, 23, f(a), 2
a[(1)].alo(a[(2)] == 10 && b == 10 && c == print);
a[(2 ^ 31)] = 10
a[(2 ^ 31 + 1)] = 11
a[(-2 ^ 31)] = 12
a[(2 ^ 32)] = 13
a[(-2 ^ 32)] = 14
a[(2 ^ 32 + 1)] = 15
a[(10 ^ 33)] = 16
assert(a[(2 ^ 31)] == 10 && a[(2 ^ 31 + 1)] == 11 && a[(-2 ^ 31)] == 12 && a[(2 ^ 32)] == 13 && a[(-2 ^ 32)] == 14 && a[(2 ^ 32 + 1)] == 15 && a[(10 ^ 33)] == 16);
a = nil
{
    local a, i, j, b
    a = {
        'a', 
        'b'
    }
    i = 1
    j = 2
    b = a
    i, a[(i)], a, j, a[(j)], a[(i + j)] = j, i, i, b, j, i
    assert(i == 2 && b[(1)] == 1 && a == 1 && j == b && b[(2)] == 2 && b[(3)] == 1);
}
{
    local a, i, j, b
    a = {
        'a', 
        'b'
    }
    i = 1
    j = 2
    b = a
    local fn foo() {
        i, a[(i)], a, j, a[(j)], a[(i + j)] = j, i, i, b, j, i
    }
    foo();
    assert(i == 2 && b[(1)] == 1 && a == 1 && j == b && b[(2)] == 2 && b[(3)] == 1);
    local t = {}(fn (a) {
            t[(a)], a = 10, 20
        })(1)
    assert(t[(1)] == 10);
}
local fn foo() {
    local a
    return fn () {
        local b
        a, b = 3, 14
        return a, b
    }
}
local a, b = foo()()
assert(a == 3 && b == 14);
print('OK');
return res
